### ISSUE TITLE
Feature request: IOperation support for queries

### ISSUE THREAD

**Issue by bkoelman** [dotnet/roslyn#17838]
Feeding the next code sample to an `OperationWalker` hits nothing. Please consider `IOperation` support for queries.

```csharp
var query =
    from ch in text
    where ch >= 'a' && ch <= 'z'
    select char.ToUpperInvariant(ch);
```

**Comment by jinujoseph** [comment]
cc/ @genlu

**Comment by genlu** [comment]
LINQ was left out with the intention to support them in future, when we made the effort to release IOperation last year. We need to discuss again whether it should be in V1 or not.

**Comment by mavasani** [comment]
Note that not implementing this completely means that our current implementation makes no callbacks for anything inside a query - parameter/field/local references, invocations, etc.

I don't think that is acceptable for **any** new language feature. We need to take one of the following approaches:

1. Ensure that we have a temporary `IOperationWithChildren` type that ensures the bound child nodes of new language features to still be part of the Operation tree, so the analyzer driver can make callbacks for them
2. Design an API for the language feature.

**Comment by mavasani** [comment]
I am moving this back to 15.3 to discuss in the API review meeting.

**Comment by mavasani** [comment]
I have opened https://github.com/dotnet/roslyn/issues/18968 to track the work to add `IOperationWithChildren` internal type as a workaround for NYI features.

**Comment by jinujoseph** [comment]
**Design Meeting Decision**

For now lets unblock users with the workaround defined in #18968 . we need to come back and discuss about what the API for queries looks like.

**Comment by jinujoseph** [comment]
we need to collect a lot of information here so that we can then get together to discuss how we want this API to look like.

We need to identify differences between the languages (for example VB is semantic and C# is syntactic).  We need to enumerate all the different types of queries.  We need to understand all the strange VB cases.  We need to get info on transparent identifiers and the like.

Basically, you needs to go read up a lot on these guys and collect a lot of info as we'll likely need a lot of design around this.

**Comment by mavasani** [comment]
As we discussed in the last design meeting, we are going to keep the API simple and minimal for V1:

1. Bound trees for VB and C# both have an unrolled, slightly lowered form for the queries where the topmost BoundQueryClause node in the bound tree corresponds to the last query clause in syntax. Additionally, it's child is the BoundInvocationExpression for the underlying linq call.
2. We are **not** going to attempt to reverse engineer the bound tree to generate an Operation tree closer to source. Instead we will expose the tree in this lowered shape in V1, with enough information about each clause in the operation node.

For example, consider the following source code:
```
using System.Linq;

class Program
{
    static void Main(string[] args)
    {
        var x = from arg in args
                select arg;
    }
}
```

Current operation tree:
```
IOperation:  (OperationKind.None) (Syntax: 'from arg in ... select arg')
  Children(1):
      IOperation:  (OperationKind.None) (Syntax: 'select arg')
        Children(1):
            IInvocationExpression (System.Collections.Generic.IEnumerable<System.String> System.Linq.Enumerable.Select<System.String, System.String>(this System.Collections.Generic.IEnumerable<System.String> source, System.Func<System.String, System.String> selector)) (OperationKind.InvocationExpression, Type: System.Collections.Generic.IEnumerable<System.String>) (Syntax: 'select arg')
              Instance Receiver: null
              Arguments(2):
                  IArgument (ArgumentKind.Explicit, Matching Parameter: source) (OperationKind.Argument) (Syntax: 'from arg in args')
                    IConversionExpression (ConversionKind.Cast, Implicit) (OperationKind.ConversionExpression, Type: System.Collections.Generic.IEnumerable<System.String>) (Syntax: 'from arg in args')
                      IOperation:  (OperationKind.None) (Syntax: 'from arg in args')
                        Children(1):
                            IParameterReferenceExpression: args (OperationKind.ParameterReferenceExpression, Type: System.String[]) (Syntax: 'args')
                    InConversion: null
                    OutConversion: null
                  IArgument (ArgumentKind.Explicit, Matching Parameter: selector) (OperationKind.Argument) (Syntax: 'arg')
                    IConversionExpression (ConversionKind.CSharp, Implicit) (OperationKind.ConversionExpression, Type: System.Func<System.String, System.String>) (Syntax: 'arg')
                      ILambdaExpression (Signature: lambda expression) (OperationKind.LambdaExpression, Type: null) (Syntax: 'arg')
                        IBlockStatement (1 statements) (OperationKind.BlockStatement) (Syntax: 'arg')
                          IReturnStatement (OperationKind.ReturnStatement) (Syntax: 'arg')
                            IOperation:  (OperationKind.None) (Syntax: 'arg')
                    InConversion: null
                    OutConversion: null
```

Proposed Operation tree:
```
IQueryExpressionOperation (OperationKind.QueryExpression) (Syntax: 'from arg in ... select arg')
  LastClause:
      IQueryClauseOperation (QueryClauseKind.SelectClause) (OperationKind.QueryClause) (Syntax: 'select arg')
        UnderlyingExpression:
            IInvocationExpression (System.Collections.Generic.IEnumerable<System.String> System.Linq.Enumerable.Select<System.String, System.String>(this System.Collections.Generic.IEnumerable<System.String> source, System.Func<System.String, System.String> selector)) (OperationKind.InvocationExpression, Type: System.Collections.Generic.IEnumerable<System.String>) (Syntax: 'select arg')
              Instance Receiver: null
              Arguments(2):
                  IArgument (ArgumentKind.Explicit, Matching Parameter: source) (OperationKind.Argument) (Syntax: 'from arg in args')
                    IConversionExpression (ConversionKind.Cast, Implicit) (OperationKind.ConversionExpression, Type: System.Collections.Generic.IEnumerable<System.String>) (Syntax: 'from arg in args')
                       IQueryClauseOperation (QueryClauseKind.FromClause) (OperationKind.QueryClause)  (Syntax: 'from arg in args')
                        RangeVariables(1):
                          Symbol: arg
                        UnderlyingExpression:
                            IParameterReferenceExpression: args (OperationKind.ParameterReferenceExpression, Type: System.String[]) (Syntax: 'args')
                    InConversion: null
                    OutConversion: null
                  IArgument (ArgumentKind.Explicit, Matching Parameter: selector) (OperationKind.Argument) (Syntax: 'arg')
                    IConversionExpression (ConversionKind.CSharp, Implicit) (OperationKind.ConversionExpression, Type: System.Func<System.String, System.String>) (Syntax: 'arg')
                      ILambdaExpression (Signature: lambda expression) (OperationKind.LambdaExpression, Type: null) (Syntax: 'arg')
                        IBlockStatement (1 statements) (OperationKind.BlockStatement) (Syntax: 'arg')
                          IReturnStatement (OperationKind.ReturnStatement) (Syntax: 'arg')
                            IOperation:  (OperationKind.None) (Syntax: 'arg')
                    InConversion: null
                    OutConversion: null
```

API proposal:

```
namespace Microsoft.CodeAnalysis.Semantics
{
    /// <summary>
    /// Kinds of query clauses.
    /// </summary>
    public enum QueryClauseKind
    {
        None = 0x0,

        /// <summary>
        /// Indicates a from clause in C# and VB.
        /// </summary>
        FromClause = 0x1,
        /// <summary>
        /// Indicates a select clause in C# and VB.
        /// </summary>
        SelectClause = 0x2,
        /// <summary>
        /// Indicates a where clause in C# and VB.
        /// </summary>
        WhereClause = 0x3,
        /// <summary>
        /// Indicates a let clause in C# and VB.
        /// </summary>
        LetClause = 0x4,
        /// <summary>
        /// Indicates an orderby clause in C# and VB.
        /// </summary>
        OrderByClause = 0x5,
        /// <summary>
        /// Indicates a groupby clause in C# and VB.
        /// </summary>
        GroupByClause = 0x6,
        /// <summary>
        /// Indicates a group join clause in VB.
        /// </summary>
        GroupJoinClause = 0x7,
        /// <summary>
        /// Indicates a join clause in C# and VB.
        /// </summary>
        JoinClause = 0x8,
        /// <summary>
        /// Indicates a join into clause in C#.
        /// </summary>
        JoinIntoClause = 0x9,
        /// <summary>
        /// Indicates a distinct clause in VB.
        /// </summary>
        DistinctClause = 0xa,
        /// <summary>
        /// Indicates an aggregate clause in VB.
        /// </summary>
        AggregateClause = 0xb,
        /// <summary>
        /// Indicates a skip clause in VB.
        /// </summary>
        SkipClause = 0xc,
        /// <summary>
        /// Indicates a skip while clause in VB.
        /// </summary>
        SkipWhileClause = 0xd,
        /// <summary>
        /// Indicates a take clause in VB.
        /// </summary>
        TakeClause = 0xe,
        /// <summary>
        /// Indicates a take while clause in VB.
        /// </summary>
        TakeWhileClause = 0xf,
    }

    /// <summary>
    /// Represents a query expression in C# or VB.
    /// </summary>
    /// <remarks>
    /// This interface is reserved for implementation by its associated APIs. We reserve the right to
    /// change it in the future.
    /// </remarks>
    public interface IQueryExpression : IOperation
    {
        /// <summary>
        /// Last query clause in the unrolled query expression.
        /// For example, for the query expression "from x in set where x.Name != null select x.Name", the select clause is the last clause of the unrolled query expression,
        /// with the where clause as one of its descendant, and the from clause as the descendant of the where clause.
        /// </summary>
        IQueryClause LastClause { get; }
    }
}
```

We can have 2 different API designs for `IQueryClause` hierarchy.

1.
```
/// <summary>
    /// Represents a query clause in C# or VB.
    /// </summary>
    /// <remarks>
    /// This interface is reserved for implementation by its associated APIs. We reserve the right to
    /// change it in the future.
    /// </remarks>
    public interface IQueryClause : IOperation
    {
        /// <summary>
        /// <see cref="QueryClauseKind"/> of the clause.
        /// </summary>
        QueryClauseKind ClauseKind { get; }
        /// <summary>
        /// Underlying source expression for the query clause.
        /// </summary>
        IOperation UnderlyingExpression { get; }
    }

    public interface ISelectQueryClause: IQueryClause
    {

    }

    public interface IFromQueryClause: IQueryClause
    {
        /// <summary>
        /// Range variables declared by the query clause, if any.
        /// C# can have at most one range variable declaration, while VB can have multiple of them.
        /// </summary>
        ImmutableArray<IRangeVariableSymbol> RangeVariables { get; }
    }
```

2.
```
    /// <summary>
    /// Represents a query clause in C# or VB.
    /// </summary>
    /// <remarks>
    /// This interface is reserved for implementation by its associated APIs. We reserve the right to
    /// change it in the future.
    /// </remarks>
    public interface IQueryClause : IOperation
    {
        /// <summary>
        /// <see cref="QueryClauseKind"/> of the clause.
        /// </summary>
        QueryClauseKind ClauseKind { get; }
        /// <summary>
        /// Underlying source expression for the query clause.
        /// </summary>
        IOperation UnderlyingExpression { get; }
        /// <summary>
        /// Range variables declared by the query clause, if any.
        /// C# can have at most one range variable declaration, while VB can have multiple of them.
        /// </summary>
        ImmutableArray<IRangeVariableSymbol> RangeVariables { get; }
    }
```

I prefer (1) because:
1. Each clause kind is strongly typed and can be easily extended in future to have different specific IOperation nodes (for example assume we add a new clause in future with more than one underlying expressions).
2. Matches other existing API shape, for example, the ICaseClause which has subtypes IDefaultCaseClause, ISingleValueCaseClause, etc.

**Comment by mavasani** [comment]
Tagging @dotnet/analyzer-ioperation

**Comment by 333fred** [comment]
I think I prefer shape 1 as well. I also definitely think we should have a single OperationKind for Queries, and have a separate kind for each different type of query. That will help anyone who actually uses the API understand what types of queries can actually exist.

**Comment by CyrusNajmabadi** [comment]
> public interface IQueryClause : IOperation

Should IQueryClause actually be an IOperation?

**Comment by CyrusNajmabadi** [comment]
> public interface IFromQueryClause

This is interesting as you expose range variables here... but those range variables will never be used elsewhere in the actually rewritten query, right?

**Comment by CyrusNajmabadi** [comment]
> public interface IQueryClause : ... UnderlyingExpression ....

I might call this "LoweredExpression".  But the name is not that important to me.

I like this pattern a lot.  Because of this it sounds like we *could* in teh future then flesh out these clauses to contain more actual information.  For example, an ISelectClause could have the lowered expression (the invocation of .Select, with the lambda), while also *in the future* having an IOperation to represent the Expression that is much closer to source.  This could then be added in the future without needing any new APIs.

**Comment by CyrusNajmabadi** [comment]
> I prefer (1) because:

Me too.

**Comment by CyrusNajmabadi** [comment]
> We chose the first approach for different kind of case clauses, but I feel the latter is a better API.

I prefer the first form.  I don't see the need to have a single QueryClause kind.  Instead, there is just a IQueryClause root interface, with specific subtypes and specific normal operation kinds for each of those subtypes.

**Comment by CyrusNajmabadi** [comment]
I *really* like this.  I thought we woudl be going with an approach where alll you saw were the IInvocationExpressions in the tree.  I really like the intermediary IQueryExpression and IQueryClauses.  

Note: we have to define what is the right ting to do when someone says "give me the IOperation for a node".  Clearly for the Query and the Claues you should get the IQueryExpression and IQueryClause.  However, for something like "select x > 2", if you ask for the IOperation for "x > 2" we need to make sure you get the operation representing hte binary expression, and not some clause, or lambda (IMO).

**Comment by mavasani** [comment]
> I prefer the first form. I don't see the need to have a single QueryClause kind. Instead, there is just a IQueryClause root interface, with specific subtypes and specific normal operation kinds for each of those subtypes

@CyrusNajmabadi should we do something similar for case clauses? We have different OperationKinds for each case clause kind and then we also have a CaseClauseKind enum which is exposed on ICaseClause.

**Comment by jinujoseph** [comment]
**Design Team Decision**
We approve the first approach.
We should have separate IQueryClause.ClauseKind which share the same OperationKind.QueryClause
We probably don't need rangevariable here.

**Comment by mavasani** [comment]
See https://github.com/dotnet/roslyn/pull/21356#issuecomment-325413081 for discussion

**Comment by CyrusNajmabadi** [comment]
Design decisions: 
Vastly reduce the IQueryExpression API.  Do not expose clauses.  Do not find a way to map from a query clause node to anything.  Rename IQueryExpression to something like ITranslatedQueryExpression (name TBD).

Figure out engineering way to "lower" the current VB query tree so that query lambdas are rewritten to regular lambdas.  This will ensure that range variables are rewritten into lambda parameter access.  C# and VB trees should look more similar.

--

In future (potentially based on customer demand) we can supply a higher level Query API that more closely matches the original source (along with things like range variables and the like).

**Comment by mavasani** [comment]
Removing the Urgency-Now label and marking as 15.later - we have basic support for translated query expressions now, which matches our v1 scope. Keeping this issue open to track richer query APIs for v2.

**Comment by mavasani** [comment]
https://github.com/dotnet/roslyn/issues/28098 tracks exposing richer IOperation API for queries. This issue was closed as we do expose an `ITranslatedQueryOperation` for queries.
